File: Answer/1137.第-n-个泰波那契数.cs
/*
 * @lc app=leetcode.cn id=1137 lang=csharp
 *
 * [1137] 第 N 个泰波那契数
 */

// @lc code=start
public class Solution {
    public static int[] T = new int[38];

    public int Tribonacci(int n) {
        if(n == 0) return 0;
        T[0] = 0;
        T[1] = 1;
        T[2] = 1;
        if(T[n] != 0) return T[n];
        T[n] = GetTribonacci(n);
        return T[n];
    }

    public int GetTribonacci(int i)
    {
        if(i == 0) return 0;
        if(T[i] != 0) return T[i];
        T[i] = GetTribonacci(i - 1) + GetTribonacci(i - 2) + GetTribonacci(i - 3);
        return T[i];
    }
}
// @lc code=end

File: Answer/1137.第-n-个泰波那契数.py
#
# @lc app=leetcode.cn id=1137 lang=python3
#
# [1137] 第 N 个泰波那契数
#

# @lc code=start
class Solution:
    def tribonacci(self, n: int) -> int:
        # dp[n] = dp[n-1] + dp[n-2] + dp[n-3]
        if n == 0:
            return 0
        if n == 1:
            return 1
        if n == 2:
            return 1
        dp = []
        dp.append(0)
        dp.append(1)
        dp.append(1)
        for i in range(3, n + 1):
            dp.append(dp[i - 1] + dp[i - 2] + dp[i - 3])
        return dp[n]
        
# @lc code=end

